File: frontend/src/types/workflowMonitor.ts
export interface MonitorTemplateListItem {
  id: string
  name: string
  version: string
  scope: 'user' | 'admin' | string
  description?: string
}

export interface MonitorFilter {
  key: string
  label: string
  type: 'select' | 'flow_select' | 'search' | string
  default?: unknown
  placeholder?: string
  options?: Array<{ label: string; value: string }>
}

export interface MonitorWidgetLayout {
  x: number
  y: number
  w: number
  h: number
}

export interface MonitorTableColumn {
  key: string
  label: string
  width?: number
  renderer?: 'status_badge' | 'progress_bar' | 'relative_time' | string
  action?: 'open_task' | 'open_report' | 'open_logs' | string
}

export interface MonitorWidget {
  id: string
  type: 'metric' | 'task_table' | 'event_stream' | 'error_list' | 'rule_timeline' | 'worker_status' | string
  title: string
  data_source: 'summary' | 'tasks' | 'events' | 'errors' | 'task_events' | 'workers' | string
  query?: Record<string, unknown>
  style?: Record<string, unknown>
  layout: Record<string, MonitorWidgetLayout>
  columns?: MonitorTableColumn[]
  options?: Record<string, unknown>
  visible?: { roles?: string[] }
}

export interface MonitorTemplate {
  id: string
  name: string
  version: string
  description?: string
  scope: 'user' | 'admin' | string
  refresh?: {
    mode?: 'websocket' | 'polling' | string
    fallback_interval_seconds?: number
  }
  permissions?: Record<string, unknown>
  filters: MonitorFilter[]
  layout: Record<string, unknown>
  widgets: MonitorWidget[]
}

export interface WorkflowMonitorSummary {
  running_count: number
  failed_today: number
  avg_running_progress: number
  stale_task_count: number
  warning_count_10m: number
  error_count_10m: number
  total: number
}

export interface WorkflowTaskSnapshot {
  id: string
  name: string
  flow_id: string
  user_id: string
  username?: string | null
  nickname?: string | null
  status: 'pending' | 'queued' | 'running' | 'success' | 'failed' | 'cancelled' | string
  progress: number
  progress_percent: number
  progress_details?: string
  current_rule?: string
  current_job_id?: number | null
  last_event_at?: string | null
  last_error?: string
  created_at?: string | null
  started_at?: string | null
  finished_at?: string | null
}

export interface WorkflowEvent {
  schema_version?: string
  task_id: string
  flow_id?: string | null
  user_id?: string | null
  project_name?: string | null
  timestamp?: string | null
  timestamp_ns?: string | null
  level: 'debug' | 'info' | 'warning' | 'error' | 'critical' | string
  source: string
  message: string
  caller?: string | null
  snakemake?: {
    rule?: string | null
    job_id?: number | null
    event_type?: string | null
    shell_command?: string | null
    progress_percent?: number | null
    progress_details?: string | null
  }
  runtime?: Record<string, unknown>
  raw?: Record<string, unknown>
}

export interface WorkflowOverviewResponse {
  summary: WorkflowMonitorSummary
  tasks: WorkflowTaskSnapshot[]
  events: WorkflowEvent[]
  errors: WorkflowEvent[]
}

export type WorkflowMonitorWsMessage =
  | { type: 'event'; event: WorkflowEvent }
  | { type: 'task_snapshot'; task: WorkflowTaskSnapshot }
  | { type: 'summary'; summary: WorkflowMonitorSummary }
  | { type: 'error'; detail: string }
  | { type: 'pong' }

File: src/omichub/application/schemas/workflow_monitor.py
"""Workflow monitor DTOs."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from omichub.application.schemas.base import OmicsHubBaseSchema


class MonitorTemplateListItem(OmicsHubBaseSchema):
    id: str
    name: str
    version: str = "1.0.0"
    scope: str = "user"
    description: str = ""


class MonitorTemplateListResponse(OmicsHubBaseSchema):
    items: list[MonitorTemplateListItem]
    total: int


class MonitorTemplateResponse(OmicsHubBaseSchema):
    id: str
    name: str
    version: str = "1.0.0"
    description: str = ""
    scope: str = "user"
    refresh: dict[str, Any] = {}
    permissions: dict[str, Any] = {}
    filters: list[dict[str, Any]] = []
    layout: dict[str, Any] = {}
    widgets: list[dict[str, Any]] = []


class WorkflowMonitorSummary(OmicsHubBaseSchema):
    running_count: int = 0
    failed_today: int = 0
    avg_running_progress: float = 0.0
    stale_task_count: int = 0
    warning_count_10m: int = 0
    error_count_10m: int = 0
    total: int = 0


class WorkflowTaskSnapshot(OmicsHubBaseSchema):
    id: str
    name: str
    flow_id: str
    user_id: str
    username: str | None = None
    nickname: str | None = None
    status: str
    progress: float
    progress_percent: float
    progress_details: str = ""
    current_rule: str = ""
    current_job_id: int | None = None
    last_event_at: str | None = None
    last_error: str = ""
    created_at: str | None = None
    started_at: str | None = None
    finished_at: str | None = None


class WorkflowEvent(OmicsHubBaseSchema):
    schema_version: str = "omichub.workflow_event.v1"
    task_id: str
    flow_id: str | None = None
    user_id: str | None = None
    project_name: str | None = None
    timestamp: str | None = None
    timestamp_ns: str | None = None
    level: str = "info"
    source: str = "snakemake"
    message: str
    caller: str | None = None
    snakemake: dict[str, Any] = {}
    runtime: dict[str, Any] = {}
    raw: dict[str, Any] = {}


class WorkflowOverviewResponse(OmicsHubBaseSchema):
    summary: WorkflowMonitorSummary
    tasks: list[WorkflowTaskSnapshot]
    events: list[WorkflowEvent]
    errors: list[WorkflowEvent]


class WorkflowEventListResponse(OmicsHubBaseSchema):
    items: list[WorkflowEvent]
    total: int


class WorkflowMonitorWsMessage(OmicsHubBaseSchema):
    type: Literal["event", "task_snapshot", "summary", "error", "pong"]
    event: WorkflowEvent | None = None
    task: WorkflowTaskSnapshot | None = None
    summary: WorkflowMonitorSummary | None = None
    detail: str | None = None


def utc_now_iso() -> str:
    return datetime.utcnow().isoformat() + "Z"
